File: src/bootstrap.py
"""Default wiring for application ports and infrastructure adapters."""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path

from src.application.ports import (
    ClusterConnector,
    ContextSwitcher,
    InventoryCatalog,
    InventoryRefresher,
    StatusReader,
    TunnelManager,
)
from src.infrastructure.adapters.argocd_connector import LocalArgocdConnector
from src.infrastructure.adapters.cluster_connector import LocalClusterConnector
from src.infrastructure.adapters.context_switcher import KubectlContextSwitcher
from src.infrastructure.adapters.inventory_catalog import (
    GitInventoryRefresher,
    YamlInventoryCatalog,
)
from src.infrastructure.adapters.status_reader import LocalStatusReader
from src.infrastructure.adapters.tunnel_manager import LocalTunnelManager


@dataclass(frozen=True)
class ServiceContainer:
    catalog: InventoryCatalog
    refresher: InventoryRefresher
    connector: ClusterConnector
    switcher: ContextSwitcher
    status_reader: StatusReader
    tunnel_manager: TunnelManager


def build_service_container(state_dir: Path | None = None) -> ServiceContainer:
    return ServiceContainer(
        catalog=YamlInventoryCatalog(),
        refresher=GitInventoryRefresher(),
        connector=LocalClusterConnector(argocd_connector=LocalArgocdConnector()),
        switcher=KubectlContextSwitcher(),
        status_reader=LocalStatusReader() if state_dir is None else LocalStatusReader(state_dir),
        tunnel_manager=LocalTunnelManager(),
    )

File: internal/bootstrap/bootstrap.go
package bootstrap

import (
	"github.com/systemframe/k3ctx/internal/domain"
	"github.com/systemframe/k3ctx/internal/infrastructure"
)

// ServiceContainer holds infrastructure adapters consumed by CLI commands.
// Observability connectors (ArgoCD, Alertmanager, VictoriaMetrics) are
// composed into ClusterConnector inside Build and not exposed here, since
// no CLI command consumes them directly.
type ServiceContainer struct {
	Catalog     domain.InventoryCatalog
	Switcher    domain.ContextSwitcher
	Status      domain.StatusReader
	Tunnels     domain.TunnelManager
	Reconnector domain.TunnelReconnector
	Connector   domain.ClusterConnector
	Exec        domain.ClusterExec
	Preflight   domain.NetBirdPreflightChecker
}

// Build creates a ServiceContainer wired to the NetBird catalog.
func Build(cfg domain.EffectiveConfig) ServiceContainer {
	connector := infrastructure.NewLocalClusterConnector(
		infrastructure.NewLocalArgocdConnector(),
		infrastructure.NewLocalAlertmanagerConnector(),
		infrastructure.NewLocalVictoriaMetricsConnector(),
	)

	return ServiceContainer{
		Catalog: infrastructure.NetBirdInventoryCatalog{
			BinPath:    cfg.NetBirdBinPath,
			HostFilter: cfg.NetBirdHostFilter,
		},
		Switcher: infrastructure.KubectlContextSwitcher{},
		Status: infrastructure.LocalStatusReader{
			PortRangeStart: cfg.PortRangeStart,
			PortRangeSize:  cfg.PortRangeSize,
		},
		Tunnels:     infrastructure.LocalTunnelManager{},
		Reconnector: infrastructure.LocalTunnelManager{},
		Connector:   connector,
		Exec:        infrastructure.LocalClusterExec{},
		Preflight:   infrastructure.NewNetBirdPreflightChecker(cfg.NetBirdBinPath),
	}
}
